thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/ThirdEyeAnomalyApplication.java

```diff
@@ -169,7 +158,34 @@ public void stop() throws Exception {
         if (config.isClassifier()) {
           classificationJobScheduler.shutdown();
         }
+        if (config.isPinotProxy()) {
+          // Do nothing
+        }
       }
     });
   }
+
+  private void initAnomalyFunctionFactory(String functoinConfigPath) {
+    if (anomalyFunctionFactory == null) {
```

Comment:
why is this null-check needed? this code path should execute exactly once

Reply:
It could be invoked multiple times when a controller has multiple rules, e.g., worker and alerter.

We didn't have this check and the second initialization overwrites the previous value. This helper function ensures that a factory instance is reused.